### PR TITLE
Error handling and restarting in optimization

### PR DESCRIPTION
**Pull request by esantorella**
## Motivation

Several issues were discussed in #1227:
- `scipy.optimize.minimize` and functions that call it can fail without warnings or errors.
- We allow use of `sample_around_best` with equality constraints even though it likely would not produce valid solutions.

Changes in this PR:
- Issue a warning in `gen_candidates_scipy` if Scipy reports that optimization failed.
- Raise an `UnsupportedError` in `gen_batch_inital_conditions` if the user tries to use the `"sample around best"` option in combination with equality constraints
- `optimize_acqf` catches optimization warnings from `gen_candidates_scipy`. If there are user-provided `batch_initial_conditions`, it raises a warning and suggests trying different initial conditions. Otherwise, it generates new initial conditions and tries again, generating another warning if it didn't work even on the second try.

I'd be curious to hear if people think that re-trying with new initial conditions could create a performance regression. That could happen if Scipy has been overly aggressive in tagging optimization runs as "failed" when they were actually good enough.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- Test in `test_gen.py` for the optimization warning from `gen_candidates_scipy`
- Test that the appropriate `UnsupportedError` is raised in `gen_batch_initial_conditions`
Several tests that the appropriate warnings are raised and appropriate re-try happens in `optimize_acqf`:
- A case where it fails due to bad user-provided ICs
- A case where it fails due to bad ICs due to a bad seed, then succeeds on retry
- A case where it fails due to bad ICs due to a bad seed and fails again on retry

### CODE REVIEW

File: botorch/generation/gen.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import numpy as np
 import torch
 from botorch.acquisition import AcquisitionFunction
+from botorch.exceptions.warnings import OptimizationWarning
```
Comment (saitcakmak) [inline botorch/generation/gen.py:20]: There's an issue with botorch warnings getting filtered out in some cases (https://github.com/pytorch/botorch/blob/main/botorch/settings.py#L58). See #1169 for relevant issue & #1170, #1178 for fixes. Let's not use them in critical cases such as this one. `RuntimeWarning` might be a better alternative.

File: botorch/optim/optimize.py [hunk 2]
```diff
@@ -239,11 +249,60 @@ def optimize_acqf(
...
-        batch_candidates_list.append(batch_candidates_curr)
-        batch_acq_values_list.append(batch_acq_values_curr)
-        logger.info(f"Generated candidate batch {i+1} of {len(batched_ics)}.")
-    batch_candidates = torch.cat(batch_candidates_list)
-    batch_acq_values = torch.cat(batch_acq_values_list)
+
+        for i, batched_ics_ in enumerate(batched_ics):
+            # optimize using random restart optimization
+            with warnings.catch_warnings(record=True) as ws:
+                warnings.simplefilter("always", category=OptimizationWarning)
```
Comment (saitcakmak) [inline botorch/optim/optimize.py:256]: This is only effective within the context. So it'll let you catch the ones raised in `gen_candidates_scipy` here.

File: botorch/optim/optimize.py [hunk 3]
```diff
@@ -239,11 +249,60 @@ def optimize_acqf(
...
+        batch_candidates = torch.cat(batch_candidates_list)
+        batch_acq_values = torch.cat(batch_acq_values_list)
+        return batch_candidates, batch_acq_values, opt_warnings
+
+    batch_candidates, batch_acq_values, ws = _optimize_batch_candidates()
+
+    optimization_warning_raised = any(
+        (issubclass(w.category, OptimizationWarning) for w in ws)
+    )
+    if optimization_warning_raised:
```
Comment (saitcakmak) [inline botorch/optim/optimize.py:274]: But the warnings in here may not get raised.

**Review by saitcakmak — APPROVED** [review]
Thanks! A series of comments in line regarding the behavior of BoTorch warnings. 

Thoughts on raising a warning vs an error if everything failed? We use warnings in `fit_gpytorch_model`, so having warnings here is consistent with that. It will let you keep the optimization going, though it may end up with a pathological behavior of never evaluating feasible points. No strong preference either way. cc @Balandat

**Comment by Balandat** [comment]
> Thoughts on raising a warning vs an error if everything failed? We use warnings in fit_gpytorch_model, so having warnings here is consistent with that. It will let you keep the optimization going, though it may end up with a pathological behavior of never evaluating feasible points. No strong preference either way.

Let's stick with warnings for now. There may be situations in which every once in a while the acquisition optimization fails but most of the time everything works fine. In those cases a warning will, well, just warn, rather than error and potentially break a bunch of downstream code. We can reconsider this later, but for now let's take one step at a time :)

**Comment by esantorella** [comment]
> Let's stick with warnings for now. There may be situations in which every once in a while the acquisition optimization fails but most of the time everything works fine. In those cases a warning will, well, just warn, rather than error and potentially break a bunch of downstream code. We can reconsider this later, but for now let's take one step at a time :)

On a similar note, are there any concerns about restarting with new initial conditions if `minimize` fails on the first try? If `minimize` is being overly sensitive in saying it failed (perhaps if a tolerance has been set too small), then could this create a substantial performance drag?

**Comment by Balandat** [comment]
> On a similar note, are there any concerns about restarting with new initial conditions if minimize fails on the first try? If minimize is being overly sensitive in saying it failed (perhaps if a tolerance has been set too small), then could this create a substantial performance drag?

I am not sure, we'd have to look at a range of cases to get a sense for how prevalent these optimization failures of the acquisition function are. Maybe we can use @mpolson64's benchmarking setup to do that?

File: botorch/generation/gen.py [hunk 4]
```diff
@@ -212,18 +212,22 @@ def f(x):
...
 
-    if "success" not in res.keys():
-        warnings.warn(
-            "Optimization failed within `scipy.optimize.minimize` with no "
-            "status returned to `res.`",
-            OptimizationWarning,
-        )
+    if "success" not in res.keys() or "status" not in res.keys():
+        with warnings.catch_warnings():
+            warnings.simplefilter("always", category=OptimizationWarning)
```
Comment (saitcakmak) [inline botorch/generation/gen.py:217]: I'm not a fan of having these two lines repeated so many times in the code. Probably not a concern here, but this also prevents the user from suppressing these warnings if they wanted to.

File: botorch/optim/optimize.py [hunk 5]
```diff
@@ -272,24 +272,23 @@ def _optimize_batch_candidates() -> Tuple[Tensor, Tensor, List[Warning]]:
...
+            " Suggested remediation: Try again with different "
+            "`batch_initial_conditions`, or don't provide `batch_initial_conditions.`"
+            if initial_conditions_provided
+            else "Optimization failed in `gen_candidates_scipy` with the following "
+            f"warning(s):\n{[w.message for w in ws]}\nTrying again with a new "
+            "set of initial conditions."
+        )
+        with warnings.catch_warnings():
+            warnings.simplefilter("always", category=OptimizationWarning)
+            warnings.warn(first_warn_msg, OptimizationWarning)
```
Comment (saitcakmak) [inline botorch/optim/optimize.py:289]: How about changing the type of this warning and the one below to something like `RuntimeWarning` rather than using the context & filter?

File: botorch/generation/gen.py (hunk 4 above)
Comment (esantorella) [reply botorch/generation/gen.py:217]: I don't think it does -- I suppressed these in a few tests where they weren't relevant.

**Comment by esantorella** [comment]
> > On a similar note, are there any concerns about restarting with new initial conditions if minimize fails on the first try? If minimize is being overly sensitive in saying it failed (perhaps if a tolerance has been set too small), then could this create a substantial performance drag?
> 
> I am not sure, we'd have to look at a range of cases to get a sense for how prevalent these optimization failures of the acquisition function are. Maybe we can use @mpolson64's benchmarking setup to do that?

That sounds good to me. Any opinion on testing it through the benchmarking setup vs. just landing this and waiting to see what happens in the nightly run?

File: botorch/generation/gen.py (hunk 4 above)
Comment (saitcakmak) [reply botorch/generation/gen.py:217]: Hmm, this still raises the warning for me:
```
import warnings

warnings.simplefilter("ignore")
with warnings.catch_warnings():
    warnings.simplefilter("always")
    warnings.warn("dummy warning", RuntimeWarning)
```

**Comment by saitcakmak** [comment]
> That sounds good to me. Any opinion on testing it through the benchmarking setup vs. just landing this and waiting to see what happens in the nightly run?

I think this is benign enough that we could just land it, but it doesn't hurt to be cautious. It's super easy to kick off a benchmark run. I'll share the instructions offline

File: botorch/generation/gen.py (hunk 4 above)
Comment (esantorella) [reply botorch/generation/gen.py:217]: Oh, yeah, good catch. I was using `warnings.catch_warnings(record=True) as ws` to prevent the warning from being written to stdout, but it still seems to be added to `ws` regardless of any `ignore` filtering.

**Comment by Balandat** [comment]
> I think this is benign enough that we could just land it, but it doesn't hurt to be cautious. It's super easy to kick off a benchmark run. I'll share the instructions offline

Yeah there is no particular rush getting this in right away, so let's run the benchmarks to be sure.
